### PR TITLE
Migrate generate downstream to go

### PR DESCRIPTION
**Pull request by trodge**
**Release Note Template for Downstream PRs (will be copied)**

```release-note:none

```

### CODE REVIEW

File: .ci/magician/cmd/generate_downstream.go [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+	}
+
+	if err := pushCommit(upstreamRepo, scratchRepo, baseBranch, command, commitMessage, ref, gh, rnr, ctlr); err != nil {
+		fmt.Println("Error pushing commit: ", err)
+		os.Exit(1)
+	}
+}
+
+func cloneRepo(baseBranch, repo, version string, ctlr *source.Controller) (*source.Repo, *source.Repo, error) {
+	upstreamRepo := &source.Repo{
```
Comment (ScottSuarez) [inline .ci/magician/cmd/generate_downstream.go:124]:
```suggestion
	downstreamRepo := &source.Repo{
```

downstream is indicative of the relationship between these two repos.

Comment (ScottSuarez) [reply .ci/magician/cmd/generate_downstream.go:124]: I understand the original script used this terminology but in retrospect it seems incorrect.

https://github.com/GoogleCloudPlatform/magic-modules/blob/5f5f9b3fee520b56f9dae6ec164d855d9074df87/.ci/scripts/build-environment/downstream-builder/generate_downstream.sh#L8

File: .ci/magician/cmd/generate_downstream.go [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+		Name:    upstreamRepo.Name,
+		Branch:  baseBranch,
+		Owner:   "modular-magician",
+		Path:    upstreamRepo.Path,
+		Version: upstreamRepo.Version,
+	}
+	return upstreamRepo, scratchRepo, nil
+}
+
+func setConfig(rnr *exec.Runner) error {
```
Comment (ScottSuarez) [inline .ci/magician/cmd/generate_downstream.go:169]:
```suggestion
func setGitConfig(rnr *exec.Runner) error {
```

File: .ci/magician/cmd/generate_downstream.go [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+	}
+	var pullRequest github.PullRequest
+	for _, pr := range prs {
+		if pr.MergeCommitSha == ref {
+			pullRequest = pr
+			break
+		}
+	}
+
+	if command == "downstream" && upstreamRepo.Title == "terraform" {
```
Comment (ScottSuarez) [inline .ci/magician/cmd/generate_downstream.go:280]: I believe this should be moved out of the push commit function to a new function called merge/push pull request and called from the toplevel `execGenerateDownstream`. That ways pushing commit and merging a PR are clear from the toplevel command.

File: .ci/magician/cmd/generate_downstream.go [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+	if _, err := rnr.Run("git", []string{"config", "--local", "user.name", "Modular Magician"}, nil); err != nil {
+		return err
+	}
+	if _, err := rnr.Run("git", []string{"config", "--local", "user.email", "magic-modules@google.com"}, nil); err != nil {
+		return err
+	}
+	return nil
+}
+
+func branchAndCommitMessage(mmRepo, scratchRepo *source.Repo, command, ref string, rnr *exec.Runner, ctlr *source.Controller) (string, error) {
```
Comment (ScottSuarez) [inline .ci/magician/cmd/generate_downstream.go:179]: Its a bit confusing to create the scratch repo then mutate it in a sequential function. Could we merge this into the clone REPO function or create a scratch repo afterwards from the original

File: .ci/magician/cmd/generate_downstream.go [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+		fmt.Println(repo, " is no longer available.")
+		return nil, nil, nil
+	default:
+		return nil, nil, fmt.Errorf("unrecognized repo %s", repo)
+	}
+	ctlr.SetPath(upstreamRepo)
+	if err := ctlr.Clone(upstreamRepo); err != nil {
+		return nil, nil, err
+	}
+	scratchRepo := &source.Repo{
```
Comment (ScottSuarez) [inline .ci/magician/cmd/generate_downstream.go:159]: nit: its kind of hard to follow source repo and scratch repo's state since they are interconnected. Notably the branch changes on the checked out branch which then makes the other repos 'state' invalid. It's albiet a bit confusing from my perspective.

**Review by ScottSuarez — CHANGES_REQUESTED** [review]

File: .ci/magician/cmd/generate_downstream.go (hunk 5 above)
Comment (trodge) [reply .ci/magician/cmd/generate_downstream.go:172]: I'm not sure what you mean here. The scratch repo refers to the fork owned by the magician, while the downstream repo is the actual repo.

File: .ci/magician/cmd/generate_downstream.go (hunk 1 above)
Comment (trodge) [reply .ci/magician/cmd/generate_downstream.go]: Done

File: .ci/magician/cmd/generate_downstream.go (hunk 3 above)
Comment (trodge) [reply .ci/magician/cmd/generate_downstream.go]: Done

File: .ci/magician/cmd/generate_downstream.go (hunk 2 above)
Comment (trodge) [reply .ci/magician/cmd/generate_downstream.go]: Done

File: .ci/magician/cmd/generate_downstream.go (hunk 4 above)
Comment (trodge) [reply .ci/magician/cmd/generate_downstream.go]: Done

File: .ci/magician/cmd/generate_downstream.go (hunk 5 above)
Comment (ScottSuarez) [reply .ci/magician/cmd/generate_downstream.go:172]: Yes, but source controller (in my mental model) is used to clone and track a repository. It's a bit weird how we clone it and the metadata of the object says the state is on a certain branch when they are both tracking the same physical repo location.

**Review by ScottSuarez — APPROVED** [review]
